### PR TITLE
SSE-CPK support added to MLflow

### PR DESCRIPTION
**Pull request by kriscon-db**
## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

In this PR we:

- Add protobuf definitions to support the new presigned URL type
- Add azure client support for ADLS Gen2 HTTP create, append and flush operations
- Update the artifact repositories that use the current Azure blob store upload/download APIs to be able to handle the new PresignedUrl type and handle ADLS Gen2 URLs.
- Single part and multipart upload for ADLS Gen2 in the databricks artifact repository.
- Threaded encryption headers into all upload and download calls
- Unit tests

## How is this patch tested?

We have added new unit tests to unit test the repository code paths.  We have also manually tested this change on a CPK test shard and manually validated that this feature branch works.

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by jinzhang21 — COMMENTED** [review]
Replaced @dbczumar with @harupy to lessen his burden on reviews. Feel free to still take a look if you wish, @dbczumar !

File: mlflow/spark.py
Comment (liangz1) [inline mlflow/spark.py:796]: This change is not related to supporting ADLS Gen 2.
I moved these two lines after checking mlflowdbfs because mlflowdbfs doesn't need `_download_artifact_from_uri` to download the model to the driver's local filesystem. The model will be downloaded by mlflowdbfs. So without this code change, the model will be downloaded twice, which I think is unnecessarily harming the performance.

However, it looks like the `local_model_path` is needed by `_add_code_from_conf_to_system_path`, which should be called before loading the model. Could anyone help me understand in which scenario would Spark model need to use the custom code path? I looked at the test [here](https://src.dev.databricks.com/mlflow/mlflow/-/blob/tests/spark/test_spark_model_export.py?L829)  and I still don't get why we need to add the `__file__` to the system path.

If we do need to support this, we should REVERT the code change in this file.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (jinzhang21) [inline mlflow/store/artifact/databricks_artifact_repo.py:254]: Please make sure these are the only codes will be returned on authorization failure scenarios.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (jinzhang21) [inline mlflow/store/artifact/databricks_artifact_repo.py:262]: Under what circumstances does this special handling of refreshing credentials will actually work? or why does it fail in the first place that we only need to retry exactly once to resolve?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (jinzhang21) [inline mlflow/store/artifact/databricks_artifact_repo.py:294]: Shouldn't `credentials` be reset as `credentials_infos` here before retrying? What happens for the next chunk?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (jinzhang21) [inline mlflow/store/artifact/databricks_artifact_repo.py:272]: Calling the length of a chunk `offset` is really confusing. Maybe `chunk_size`?

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (liangz1) [reply mlflow/store/artifact/databricks_artifact_repo.py:262]: In my understanding, presigned urls will expire after 300s or so (maybe 900s, I don't remember the exact number), if there are many chunks and the uploading takes longer than that, the `put` will fail. In this case, getting a new presigned url for write will resolve this issue.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (liangz1) [reply mlflow/store/artifact/databricks_artifact_repo.py:294]: It seems that all the occurrences of `credentials_infos` in this function should be replaced with `credentials`.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (jinzhang21) [reply mlflow/store/artifact/databricks_artifact_repo.py:262]: OK. This makes sense to me.

Comment (jinzhang21) [reply mlflow/store/artifact/databricks_artifact_repo.py:262]: @kriscon-db Can we try to abstract away this try-except block with an explicit aim to retry with refreshed credentials? It's very repetitive w/o clear indication of its functionality. Something like `def retry_with_credential_refresh(fn, credentials):`, or make it a decorator if you want.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (kriscon-db) [reply mlflow/store/artifact/databricks_artifact_repo.py:254]: It actually looks like there are no 401's returned?  https://learn.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create#response

Reading through these codes, the one that would trigger on a potential token expiration is one of the 403's.  I'll change to limit to just 403.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (kriscon-db) [reply mlflow/store/artifact/databricks_artifact_repo.py:262]: Yep.  Let me push a fix.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (kriscon-db) [reply mlflow/store/artifact/databricks_artifact_repo.py:272]: Will rename to `cursor_position` and rename `cur_offset` to `cur_chunk_size`

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (kriscon-db) [reply mlflow/store/artifact/databricks_artifact_repo.py:294]: Will fix.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (kriscon-db) [reply mlflow/store/artifact/databricks_artifact_repo.py]: Please re-review this section.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (kriscon-db) [inline mlflow/store/artifact/databricks_artifact_repo.py:243]: To avoid the potential repeated calls to `_get_write_credential_infos`, we could have this method return the final credentials used and the calling points would overwrite their credentials with what is returned.  I am not a fan of this because it is tricky/clever code and tricky/clever code takes awhile to understand when other engineers have to read it.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (harupy) [inline mlflow/store/artifact/databricks_artifact_repo.py:293]:
```suggestion
            use_single_part_upload = False
            for idx, chunk in enumerate(yield_file_in_chunks(local_file, _AZURE_MAX_BLOCK_CHUNK_SIZE):
                cur_chunk_size = len(chunk)
                if idx == 0 and cur_chunk_size < _AZURE_MAX_BLOCK_CHUNK_SIZE:
                    use_single_part_upload = True
                self._retryable_adls_function(
                    func=patch_adls_file_upload,
                    artifact_path=artifact_path,
                    sas_url=credentials.signed_uri,
                    data=chunk,
                    position=cursor_position,
                    headers=headers,
                    is_single=use_single_part_upload,
                )
```

Can we use `enumerate` and remove `is_first_chunk`?

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (harupy) [inline tests/store/artifact/test_databricks_artifact_repo.py:378]:
```suggestion
            DATABRICKS_ARTIFACT_REPOSITORY + "._get_write_credential_infos",
            return_value=[...]
```

Nit: we can specify `return_value` when calling `mock.patch`

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (harupy) [inline tests/store/artifact/test_databricks_artifact_repo.py:402]:
```suggestion
            assert request_mock.mock_calls == [
                mock.call(
                    "put",
                    MOCK_ADLS_GEN2_SIGNED_URI + "?resource=file",
                    headers={},
                ),
                # test with azure max block size
                mock.call(
                    "patch",
                    MOCK_ADLS_GEN2_SIGNED_URI + "?action=append&position=0&flush=true",
                    data=ANY,
                    headers={},
                ),
            ]
```

to ensure the number of requests and order.

File: mlflow/azure/client.py
Comment (harupy) [inline mlflow/azure/client.py:54]: Why do we set `flush` to `true` only when we have a single chunk? What happens if we have multiple chunks and set `flush` to `true` in the final append?


```
append
append
append with flush
```

Comment (kriscon-db) [reply mlflow/azure/client.py:54]: I believe this is a shortcut that microsoft added to reduce the number of calls in a single part upload.  My understanding is this only works when the position is 0 for a single part upload.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (jinzhang21) [inline mlflow/store/artifact/databricks_artifact_repo.py:257]: `put_adls_file_creation` -> `func`?

**Review by jinzhang21 — APPROVED** [review]
LGTM on a high level. I didn't go through every piece of the code though. Please wait for @harupy 's LGTM before merging.

File: mlflow/store/artifact/databricks_artifact_repo.py
Comment (kriscon-db) [reply mlflow/store/artifact/databricks_artifact_repo.py:293]: Will change.

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (kriscon-db) [reply tests/store/artifact/test_databricks_artifact_repo.py:402]: Will change

File: tests/store/artifact/test_databricks_artifact_repo.py
Comment (kriscon-db) [reply tests/store/artifact/test_databricks_artifact_repo.py:378]: Fixing

**Review by harupy — APPROVED** [review]
LGTM!

File: mlflow/spark.py
Comment (kriscon-db) [reply mlflow/spark.py:796]: I believe this is the only open question.  cc @jinzhang21 who may have context on this?

**Comment by kriscon-db** [comment]
After talking with @liangz1, I'm reverting the spark.py change.  We can revisit this in a followup commit.

**Pull request merged**
